modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteLimit.java

```diff
@@ -99,44 +101,45 @@ public IgniteLimit(
 
     /** {@inheritDoc} */
     @Override public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
-        double rows = estimateRowCount(mq);
+        double inputRowCount = mq.getRowCount(getInput());
+
+        int lim = intFromRex(fetch);
+        int off = intFromRex(offset);
+
+        double rows = Math.min(lim + off, inputRowCount);
 
-        return planner.getCostFactory().makeCost(rows, 0, 0);
+        return planner.getCostFactory().makeCost(rows, rows * IgniteCost.ROW_PASS_THROUGH_COST, 0);
     }
 
     /** {@inheritDoc} */
     @Override public double estimateRowCount(RelMetadataQuery mq) {
-        Integer lim = intFromRex(fetch);
-        Integer off = intFromRex(offset);
+        double inputRowCount = mq.getRowCount(getInput());
 
-        if (lim == null) {
-            // If estimated rowcount is less than offset return 0
-            if (off == null)
-                return getInput().estimateRowCount(mq) * 0.01;
```

Comment:
0.01 here is due to we can`t (probably it`s possible but not implemented yet) calculate if we have DYNAMIC_PARAM or CALCULATED here, i think we need to preserve it here.

Reply:
ok, I've return this back